src/Data/Text.hs

```diff
@@ -804,7 +824,7 @@ replace needle@(Text _      _      neeLen)
                (Text repArr repOff repLen)
       haystack@(Text hayArr hayOff hayLen)
   | neeLen == 0 = emptyError "replace"
-  | L.null ixs  = haystack
+  | L.null ixs && not (null haystack) = haystack
```

Comment:
What's the reason for this change?

Reply:
If I don't have this then if the haystack is `empty` then it gets reboxed due to ww giving us a distinct heap object. 
Or in any case the corresponding unit test fails.

I'll take another look at some point and see if we can make it simpler though.